### PR TITLE
Enable ca2000 for tests and fix the leaks

### PR DESCRIPTION
**Pull request by JaynieBai**
Fixes CA2000 in tests

### Context
#9983 temporarily made the exception for test projects when enabling CA2000. 

### Changes Made
Enable CA2000 in tests.

### Testing
N/A

### CODE REVIEW

**Review by YuliiaKovalova — APPROVED** [review]

File: src/Build.UnitTests/BuildEventArgsSerialization_Tests.cs [hunk 1]
```diff
@@ -892,8 +892,10 @@ public void ReadingCorruptedStreamThrows()
                 memoryStream.SetLength(i); // pretend that the stream abruptly ends
                 memoryStream.Position = 0;
 
+#pragma warning disable CA2000 // The memory stream needs to keep the binary reader open.
```
Comment (ladipro) [inline src/Build.UnitTests/BuildEventArgsSerialization_Tests.cs:895]: nit: `BuildEventArgsReader` can be instructed to leave the reader open by setting `CloseInput = false`.

File: src/Build.UnitTests/Definition/Toolset_Tests.cs [hunk 2]
```diff
@@ -51,48 +51,39 @@ public void ToolsetCtorErrors3()
...
+                    ("C:", @"C:"),
+                    (@"C:\", @"C:\"),
+                    (@"C:\\", @"C:\"),
+                    (@"C:\foo", @"C:\foo"),
+                    (@"C:\foo\", @"C:\foo"),
+                    (@"C:\foo\\", @"C:\foo\"), // trim at most one slash
+                    (@"\\foo\share", @"\\foo\share"),
+                    (@"\\foo\share\", @"\\foo\share"),
+                    (@"\\foo\share\\", @"\\foo\share\"), // trim at most one slash
+                ];
```
Comment (ladipro) [inline src/Build.UnitTests/Definition/Toolset_Tests.cs:68]: ❤️

**Review by ladipro — APPROVED** [review]

**Review by maridematte — APPROVED** [review]

File: src/Build.UnitTests/BackEnd/TaskExecutionHost_Tests.cs [hunk 3]
```diff
@@ -1560,7 +1560,8 @@ private ProjectInstance CreateTestProject()
                 </Project>
                 ");
 
-            Project project = new Project(XmlReader.Create(new StringReader(projectFileContents)));
```
Comment (rainersigwald) [inline src/Build.UnitTests/BackEnd/TaskExecutionHost_Tests.cs:1563]: This pattern is so prevalent, I wonder if we should make a new test helper to make it smoother, something like

```c#
public class ProjectFromString : IDisposable
{
    public Project Project {get;init;}
    
    private XmlReader _reader;
    
    public ProjectFromString(string s)
    {
        _reader = XmlReader.Create(new StringReader(s));
        
        Project = new(_reader);
    }

    public void Dispose()
    {
        ((IDisposable)_reader).Dispose();
    }
}
```

And replace the existing calls with

```c#
using var projectFromString = new (projectFileContents);
Project project = projectFromString.Project;
```

? I guess the impact at every callsite is about the same, though.

Comment (JaynieBai) [reply src/Build.UnitTests/BackEnd/TaskExecutionHost_Tests.cs:1563]: Updated

File: src/Build.UnitTests/BuildEventArgsSerialization_Tests.cs (hunk 1 above)
Comment (JaynieBai) [reply src/Build.UnitTests/BuildEventArgsSerialization_Tests.cs:897]: The default CloseInput is false https://github.com/dotnet/msbuild/blob/9bea8026aad964cb36f3ec9d93bd95a941487690/src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs#L79-L82

**Comment by YuliiaKovalova** [comment]
Thank you 😊

**Pull request merged**